### ISSUE TITLE
2 members on WebSocket throw PNSE

### ISSUE THREAD

**Issue by danmosemsft** [dotnet/corefx#18905]
These are both part of NS2.0. Both show no usage in our data. Please triage to 2.0, future, or close.

```c#
        public static void RegisterPrefixes()
        {
            // The current WebRequest implementation in corefx does not support upgrading
            // web socket connections.  For now, we throw.
            throw new PlatformNotSupportedException();
        }

        [EditorBrowsable(EditorBrowsableState.Never)]
        public static WebSocket CreateClientWebSocket(Stream innerStream,
            string subProtocol, int receiveBufferSize, int sendBufferSize,
            TimeSpan keepAliveInterval, bool useZeroMaskingKey, ArraySegment<byte> internalBuffer)
        {
            // ClientWebSocket on Unix is implemented in managed code and can be constructed (internally)
            // for an arbitrary stream. We could use that implementation here, building it in to the WebSocket
            // library as well, or accessing it from the client library via reflection.  For now, we throw.
            throw new PlatformNotSupportedException();
        }
```

**Comment by stephentoub** [comment]
(RegisterPrefixes is also [EditorBrowsable(EditorBrowsableState.Never)])

**Comment by davidsh** [comment]
cc: @CIPop

**Comment by CIPop** [comment]
@weshaggard I believe that having these APIs in the Standard will only create confusion. 
Both are implementation details and documented as such on MSDN [CreateClientWebSocket](https://msdn.microsoft.com/en-us/library/system.net.websockets.websocket.createclientwebsocket(v=vs.110).aspx) and [RegisterPrefixes](https://msdn.microsoft.com/en-us/library/system.net.websockets.websocket.registerprefixes(v=vs.110).aspx)

> This API supports the .NET Framework infrastructure and is not intended to be used directly from your code.

**Comment by CIPop** [comment]
@davidsh @karelz Let's decide on a single label for WS related issues. I'm seeing that System.Net, System.Net.Sockets as well as System.Net.Http labels are used.

From all of these options I think the Sockets label is the least preferable.

**Comment by karelz** [comment]
I don't like System.Net, so System.Net.Http is IMO best candidate (minor vote) -- the question is if it will be overloaded soon.
@davidsh thoughts? cc @Priya91 @geoffkizer 

Once we decide, we can explicitly call it out in [areas list](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md#areas), so that we do not forget ;-)

**Comment by davidsh** [comment]
>I don't like System.Net, so System.Net.Http is IMO best candidate (minor vote) -- the question is if it will be overloaded soon.
@davidsh thoughts? cc @Priya91 @geoffkizer

I think of WebSockets as similar in semantics as Sockets.  So, I think it should be System.Net.Sockets label.

**Comment by Priya91** [comment]
I vote for system.net, as websocket protocol is just another application layer protocol for bidirectional communication between client and server. The RFC states that the design of the protocol does not limit websocket to http. I'm less inclined to choose Sockets area, as socket is just an ipaddress and port at layer 4  or so in OSI, whereas websocket is a protocol above that layer.

**Comment by Priya91** [comment]
For the original issue, I don't clearly understand the scenarios where one will use these APIs in their apps. Going with msdn documentation, I would make it Future. @CIPop @davidsh might have more insights.

**Comment by CIPop** [comment]
> For the original issue, I don't clearly understand the scenarios where one will use these APIs in their apps. 

@Priya91 See my comment above: we should not have these APIs available at all. By design, no application should _ever_ call them. We don't support that scenario on .Net Framework.

@weshaggard @danmosemsft Is it possible to remove them from Standard at this point?

**Comment by danmosemsft** [comment]
We do not remove members from types in net standard unless forced to to break dependencies (eg., on CAS or remoting types we don't want). So if the members cannot be implemented, PNSE is the way to go. We're trying to be pretty strict about not deferring PNSE without good reason, but of course that has to be weighed against value, usefulness, and other work you guys have. As I mentioned at the top, thes don't show up in our usage graphs.

**Comment by karelz** [comment]
If the APIs behave differently on various platforms/OSs (incl. PNSE), please document the difference here: https://github.com/dotnet/corefx/wiki/ApiCompat

**Comment by stephentoub** [comment]
FWIW, it think it would be pretty quick and easy to implement:
```C#
public static WebSocket CreateClientWebSocket(Stream innerStream,
            string subProtocol, int receiveBufferSize, int sendBufferSize,
            TimeSpan keepAliveInterval, bool useZeroMaskingKey, ArraySegment<byte> internalBuffer)
```
We basically already have that implemented here:
https://github.com/dotnet/corefx/blob/master/src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs#L39
so I expect this implementation would involve a) building ManagedWebSocket.cs into System.Net.WebSockets.dll (i.e. augment the .csproj), and b) change the method from:
```C#
throw new PlatformNotSupportedException();
```
to something like:
```C#
return CreateFromConnectedStream(
    innerStream, false, subProtocol, keepAliveInternal, receiveBufferSize, internalBuffer);
```
That ignores the `useZeroMaskingKey` and `sendBufferSize` arguments... I think ignoring the latter is fine as it's an implementation detail, but for the former we could either choose to throw a PNSE if `useZeroMaskingKey` is false or we could add support for it to ManagedWebSocket.

Then we could decide whether to continue building ManagedWebSocket into System.Net.WebSockets.Client.dll or whether to instead call CreateClientWebSocket where we currently call CreateFromConnectedStream.  While it would result in slightly larger combined IL sizes, I expect we'll want to keep System.Net.WebSockets.Client.dll as-is for 2.0, as IIRC there was a measurable impact to the virtual calls that were avoided by using the concrete sealed ManagedWebSocket type in the ClientWebSocket implementation, though we could re-test/evaluate.

Of course we could still choose to throw PNSE for CreateClientWebSocket in 2.0 even if it's feasible to implement.  And RegisterPrefixes is harder.  I expect that'll need to remain PNSE for 2.0.

**Comment by karelz** [comment]
Triage: These APIs are not called anywhere we know. MSDN says don't use them. We likely cannnot make them fully compatible with Desktop (we have no coverage). If anyone comes that it doesn't work, we won't fix them as they are by design internal / MSDN-documented not to be used.

Pending decision:
1. Remove from .NET Standard
2. Keep them, throw PNSE (document in ApiCompat)
3. Keep them, implement them as best as we can

**Comment by weshaggard** [comment]
>Remove from .NET Standard

At this point we are pretty locked down on the APIs in netstandard2.0 so it would require shiproom approval to modify it there.

**Comment by danmosemsft** [comment]
Karel as I mentioned above I do not think we have any precedent for removing from Standard any individual members  (except to break dep's). I don't think that should be on the table.

**Comment by karelz** [comment]
Thanks for details - that makes option (1) highly unlikely IMO.
Any preferences between (2) and (3)? ... I would be fine with (2) and rather fix something which we believe could have impact on real customer scenarios. Thoughts?

cc @davidsh @CIPop @DavidGoll @schen381 @geoffkizer @Priya91

**Comment by stephentoub** [comment]
Let's do (2).

**Comment by karelz** [comment]
OK, sold. Closing.
If anyone has different opinion, please speak up / reopen.

**Comment by davidsh** [comment]
Why did we close this issue?  I thought we were moving it to Future as part of 2.

**Comment by karelz** [comment]
What kind of work would it track in Future? Solution (2) is already in place, isn't it? Or am I missing some pieces of the story here?

If we need to switch to (3) based on customer feedback - either a new issue will be filed or we will reopen this one (if someone starts commenting on it prior to filing a new one)

FYI: I documented the APIs in ApiCompat wiki: https://github.com/dotnet/corefx/wiki/ApiCompat#systemnetwebsockets---registerprefixes-and-createclientwebsocket

**Comment by davidsh** [comment]
I thought choice 2 was keep it as-is (PNSE) for 2.0 but fix it in Future.  I thought choice 3 was fix it now for 2.0. I never thought choice 2 was close it and wait for customer feedback. But if that is choice 2 then ok.

**Comment by karelz** [comment]
I think that we should either fix it now (3 - if we believe it is important), or punt it (2 - due to no usage and warnings on MSDN) until we get evidence that it is used regardless the warnings and blocks adoption of .NET Core - if that happens, we can switch back to option (3), but it will be reactive, not proactive (i.e. moved to Future).
